File: freight-triage-polyglot-ledger-recovery/environment/native/src/hash/sdbm.cpp
#include "freight/hashes.h"

namespace freight {

// sdbm over raw bytes.
uint64_t hash_sdbm(const std::string& data) {
  uint32_t state = 0u;
  for (size_t i = 0; i < data.size(); ++i) {
    uint32_t byte = static_cast<uint8_t>(data[i]);
    state = byte + (state << 6) + (state << 16) - state;
  }
  return static_cast<uint64_t>(state);
}

}  // namespace freight

File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/hash/Sdbm.java
package com.freight.hash;

/** sdbm over raw bytes. */
public final class Sdbm implements HashAlgorithm {

    @Override
    public String name() {
        return "sdbm";
    }

    @Override
    public long apply(byte[] data) {
        long state = 0L;
        for (int i = 0; i < data.length; i++) {
            state = ((data[i] & 0xFF) + (state << 6) + (state << 16) - state) & 0xFFFFFFFFL;
        }
        return state;
    }
}
